### PR TITLE
feature: Downloading onebox images to local

### PR DESCRIPTION
**Pull request by vinothkannans**
Two new site settings:

1. download_onebox_images_to_local
2. download_only_http_onebox_images

### CODE REVIEW

**Comment by SamSaffron** [comment]
awesome! @ZogStriP  can you review?

**Comment by vinothkannans** [comment]
Currently it have a problem below. Which should be fixed before merging.

https://meta.discourse.org/t/download-images-for-oneboxes-as-well-if-download-images-is-set/21103/38?u=vinothkannans

File: app/jobs/regular/pull_hotlinked_images.rb [hunk 1]
```diff
@@ -89,11 +89,50 @@ def execute(args)
...
+        end
+      end
+
+      if cooked != orig_cooked
+
+        if orig_cooked.present? && cooked.blank?
+          # TODO suicide if needed, let's gather a few here first
+          Rails.logger.warn("Cooked post processor in FATAL state, bypassing. You need to urgently restart sidekiq\norig: #{orig_cooked}\ncooked: #{cooked}\npost id: #{post.id}")
+        else
+          post.update_column(:cooked, cooked)
```
Comment (tgxworld) [inline app/jobs/regular/pull_hotlinked_images.rb]: We need to make `updated_at` is updated too.

**Comment by tgxworld** [comment]
The new logic needs to be tested as well.

File: app/jobs/regular/pull_hotlinked_images.rb [hunk 2]
```diff
@@ -89,11 +89,50 @@ def execute(args)
...
+      orig_cooked = cooked.dup
+
+      extract_images_from(cooked).each do |image|
+        src = image['src']
+        src = "http:" + src if src.start_with?("//")
+        if is_valid_image_url(src)
+          if downloaded_urls[src].present?
+            url = downloaded_urls[src]
+            escaped_src = Regexp.escape(src)
+            # there are 6 ways to insert an image in a post
```
Comment (ZogStriP) [inline app/jobs/regular/pull_hotlinked_images.rb]: Not sure this comment is useful here

File: config/locales/server.en.yml [hunk 3]
```diff
@@ -878,6 +878,8 @@ en:
     download_remote_images_to_local: "Convert remote images to local images by downloading them; this prevents broken images."
     download_remote_images_threshold: "Minimum disk space necessary to download remote images locally (in percent)"
     download_remote_images_max_days_old: "Don't download remote images for posts that are more than n days old."
+    download_onebox_images_to_local: "Convert onebox images to local images by downloading them; this prevents broken images."
+    download_only_http_onebox_images: "Download onebox images only if it does not have SSL certificate."
```
Comment (ZogStriP) [inline config/locales/server.en.yml]: Not sure that site setting is really useful. It complicates the setup & code for no reasons.

File: app/jobs/regular/pull_hotlinked_images.rb [hunk 4]
```diff
@@ -89,11 +89,43 @@ def execute(args)
...
+      end
+
     end
 
     def extract_images_from(html)
       doc = Nokogiri::HTML::fragment(html)
-      doc.css("img[src]") - doc.css(".onebox-result img") - doc.css("img.avatar")
+      images = doc.css("img[src]") - doc.css("img.avatar")
+      doc.css(".onebox-body img").each do |image|
+        images -= image if image['src'].start_with?("https://")
```
Comment (ZogStriP) [inline app/jobs/regular/pull_hotlinked_images.rb]: Are we sure all oneboxed images will start with `https`?

**Comment by tgxworld** [comment]
ping @vinothkannans any updates on this PR? I think @ZogStriP  left some comments here.

**Comment by vinothkannans** [comment]
@tgxworld @ZogStriP I implemented this same feature in [kam44](https://meta.discourse.org/users/kam44)'s Discourse forum too. Then he complained as downloaded images are removed automatically (here Discourse treating it as orphans). Then I added a [commit](https://github.com/discourse/discourse/pull/4624/commits/32cda7b15d9b33628591f675694d743bde021a23) to fix that issue.

Now I am not sure about whether it is working or not. I am unable to test it. My localhost sidekiq somehow messed. May be I should write a RSpec to check downloaded images existence after orphan images cleanup job. But still I can't figure out how to write spec for image existence in this case.

Or someone else can help me by checking this PR's activity on their localhost.

**Comment by tgxworld** [comment]
> My localhost sidekiq somehow messed 

Any errors being thrown?

**Comment by vinothkannans** [comment]
@tgxworld My bad. I thought sidekiq is running on `http://localhost:3000/sidekiq/`. And I totally missed the command `bundle exec sidekiq`. Sorry I am new to many things in rails.

Now I will try to test it myself :)

**Comment by vinothkannans** [comment]
Another new PR created for this feature https://github.com/discourse/discourse/pull/4845

**Pull request closed without merging**